File: python/module-4/lesson-2/upload_traces.py
"""Upload traces with historical timestamps to LangSmith.

Uses the POST /runs REST endpoint directly (not multipart_ingest) because
the multipart endpoint enforces ±24h on start_time for creates, while the
single-run POST endpoint accepts any timestamp.
"""

import json
import uuid
import time
from datetime import datetime, timezone

import requests as http_requests
from dotenv import load_dotenv
load_dotenv()

from langsmith import Client


def parse_dt(s: str | None) -> datetime | None:
    if s is None:
        return None
    dt = datetime.fromisoformat(s)
    if dt.tzinfo is not None:
        dt = dt.replace(tzinfo=None)
    return dt


def build_dotted_order(start_time: datetime, run_id: str, parent_dotted: str | None = None) -> str:
    ts = start_time.strftime("%Y%m%dT%H%M%S%fZ")
    segment = f"{ts}{run_id}"
    if parent_dotted:
        return f"{parent_dotted}.{segment}"
    return segment


def main():
    import argparse

    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--project", default="lca-ls-project-copy", help="Target project name"
    )
    parser.add_argument(
        "--input", default="synthetic_traces.json", help="Input file path"
    )
    args = parser.parse_args()

    with open(args.input) as f:
        runs = json.load(f)

    print(f"Loaded {len(runs)} runs from {args.input}")

    # Regenerate all UUIDs
    id_map = {}
    for run in runs:
        for field in ("id", "trace_id", "parent_run_id"):
            old = run.get(field)
            if old and old not in id_map:
                id_map[old] = str(uuid.uuid4())

    # Transform runs with new IDs and proper dotted_orders
    transformed = []
    for run in runs:
        new_run = dict(run)
        new_run["id"] = id_map[run["id"]]
        new_run["trace_id"] = id_map[run["trace_id"]]
        new_run["parent_run_id"] = id_map[run["parent_run_id"]] if run["parent_run_id"] else None
        new_run["start_time"] = parse_dt(run["start_time"])
        new_run["end_time"] = parse_dt(run["end_time"])
        transformed.append(new_run)

    # Build dotted_orders: roots first, then children
    dotted_map = {}
    for r in transformed:
        if r["parent_run_id"] is None:
            r["dotted_order"] = build_dotted_order(r["start_time"], r["id"])
            dotted_map[r["id"]] = r["dotted_order"]

    for r in transformed:
        if r["parent_run_id"] is not None:
            parent_dotted = dotted_map.get(r["parent_run_id"], "")
            r["dotted_order"] = build_dotted_order(r["start_time"], r["id"], parent_dotted)
            dotted_map[r["id"]] = r["dotted_order"]

    # Upload via direct POST /runs (supports historical timestamps)
    client = Client()
    api_url = client.api_url
    headers = client._headers

    print(f"Uploading {len(transformed)} runs to project '{args.project}'...")

    # Upload roots first (parent_run_id is None), then children
    roots = [r for r in transformed if r["parent_run_id"] is None]
    children = [r for r in transformed if r["parent_run_id"] is not None]

    batch_size = 25  # smaller batches for sequential POST
    errors = 0

    for label, group in [("root", roots), ("child", children)]:
        print(f"  Uploading {len(group)} {label} runs...")
        for i in range(0, len(group), batch_size):
            batch = group[i : i + batch_size]
            for r in batch:
                payload = {
                    "id": r["id"],
                    "name": r["name"],
                    "run_type": r["run_type"],
                    "inputs": r["inputs"],
                    "outputs": r["outputs"],
                    "start_time": r["start_time"].isoformat(),
                    "end_time": r["end_time"].isoformat(),
                    "trace_id": r["trace_id"],
                    "dotted_order": r["dotted_order"],
                    "parent_run_id": r["parent_run_id"],
                    "extra": r["extra"],
                    "tags": r["tags"],
                    "error": r["error"],
                    "session_name": args.project,
                }
                resp = http_requests.post(
                    f"{api_url}/runs",
                    headers=headers,
                    json=payload,
                )
                if resp.status_code not in (200, 201, 202):
                    errors += 1
                    if errors <= 3:
                        print(f"    Error {resp.status_code}: {resp.text[:200]}")
            batch_num = i // batch_size + 1
            total_batches = (len(group) + batch_size - 1) // batch_size
            print(f"    {label} batch {batch_num}/{total_batches} ({len(batch)} runs)")

    print(f"Done! ({errors} errors)")


if __name__ == "__main__":
    main()

File: ts/module-3/lesson-2/upload_traces.ts
/**
 * Load traces.json, shift timestamps to now, regenerate IDs, and upload via RunTree.
 */

import "dotenv/config";
import * as fs from "fs";
import { parseArgs } from "util";
import { Client, RunTree } from "langsmith";
import { randomUUID } from "crypto";

function parseDt(s: string | null): Date | null {
  if (s === null || s === undefined) {
    return null;
  }
  return new Date(s);
}

async function main() {
  const { values } = parseArgs({
    options: {
      project: { type: "string", default: "default" },
      input: { type: "string", default: "synthetic_traces.json" },
    },
    strict: false,
  });

  const projectName = values.project as string;
  const inputFile = values.input as string;

  const runs: any[] = JSON.parse(fs.readFileSync(inputFile, "utf-8"));
  console.log(`Loaded ${runs.length} runs from ${inputFile}`);

  // Calculate time shift so traces appear recent
  let latest = new Date(0);
  for (const r of runs) {
    if (r.start_time) {
      const dt = parseDt(r.start_time)!;
      if (dt > latest) latest = dt;
    }
  }
  const timeDeltaMs = Date.now() - latest.getTime();
  console.log(`Shifting timestamps by: ${timeDeltaMs}ms`);

  // Build ID map (new UUIDs for time-ordering)
  const idMap: Record<string, string> = {};
  for (const run of runs) {
    for (const field of ["id", "trace_id", "parent_run_id"]) {
      const oldId = run[field];
      if (oldId && !(oldId in idMap)) {
        idMap[oldId] = randomUUID();
      }
    }
  }

  // Group runs by trace and transform
  const traces: Record<string, any[]> = {};
  for (const run of runs) {
    const traceId = run.trace_id;
    if (!traces[traceId]) traces[traceId] = [];

    const startTime = parseDt(run.start_time);
    const endTime = run.end_time ? parseDt(run.end_time) : null;

    traces[traceId].push({
      id: idMap[run.id],
      parent_run_id: idMap[run.parent_run_id] || undefined,
      name: run.name,
      run_type: run.run_type,
      inputs: run.inputs,
      outputs: run.outputs || undefined,
      error: run.error || undefined,
      extra: run.extra || undefined,
      tags: run.tags || undefined,
      start_time: startTime
        ? new Date(startTime.getTime() + timeDeltaMs)
        : undefined,
      end_time: endTime
        ? new Date(endTime.getTime() + timeDeltaMs)
        : undefined,
    });
  }

  const client = new Client();
  const traceKeys = Object.keys(traces);
  console.log(
    `Uploading ${traceKeys.length} traces to project '${projectName}'...`
  );

  for (let i = 0; i < traceKeys.length; i++) {
    const traceRuns = traces[traceKeys[i]];

    // Sort by start_time, root first (no parent)
    traceRuns.sort((a: any, b: any) => {
      const aIsChild = a.parent_run_id != null ? 1 : 0;
      const bIsChild = b.parent_run_id != null ? 1 : 0;
      if (aIsChild !== bIsChild) return aIsChild - bIsChild;
      return (
        (a.start_time?.getTime() || 0) - (b.start_time?.getTime() || 0)
      );
    });

    const treeMap: Record<string, RunTree> = {};
    let rootTree: RunTree | null = null;

    for (const run of traceRuns) {
      if (run.parent_run_id == null) {
        // Root run
        rootTree = new RunTree({
          id: run.id,
          name: run.name,
          run_type: run.run_type,
          inputs: run.inputs,
          start_time: run.start_time,
          extra: run.extra,
          tags: run.tags,
          project_name: projectName,
          client,
        });
        treeMap[run.id] = rootTree;
      } else {
        // Child run
        const parent = treeMap[run.parent_run_id];
        if (parent) {
          const child = parent.createChild({
            name: run.name,
            run_type: run.run_type,
            id: run.id,
            inputs: run.inputs,
            start_time: run.start_time,
            extra: run.extra,
            tags: run.tags,
          });
          treeMap[run.id] = child;
        }
      }
    }

    // End all runs (children first)
    for (let j = traceRuns.length - 1; j >= 0; j--) {
      const run = traceRuns[j];
      const tree = treeMap[run.id];
      if (tree) {
        tree.end({
          outputs: run.outputs,
          error: run.error,
          endTime: run.end_time,
        });
      }
    }

    if (rootTree) {
      await rootTree.postRun();
    }

    if ((i + 1) % 10 === 0) {
      console.log(`  Uploaded ${i + 1}/${traceKeys.length} traces`);
    }
  }

  // Wait for all background operations to complete
  console.log("Flushing...");
  await client.flush();
  console.log("Done!");
}

main().catch(console.error);
